### PR TITLE
Improve build times by moving various things related to correctness from Strata to StrataTest

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Partial fix for https://github.com/strata-org/Strata/issues/348
### Changes
- Move several theorems so they only build in StrataTest, improving the development speed when working on Strata
- Move all tests, identifiable by calls to `#guard` or `#guard_msg` from Strata/ to StrataTest/

Before:
```
lake build Strata 311.71s user 103.99s system 175% cpu 3:56.88 total
```
After: 
```
lake build Strata 166.28s user 70.21s system 210% cpu 1:52.31 total
```

### Testing
Refactoring that does not need any tests

### CODE REVIEW

File: Strata/DL/Lambda/LTy.lean [hunk 1]
```diff
@@ -499,16 +467,7 @@ partial def elabLTy : Lean.Syntax → MetaM Expr
...
-#guard_msgs in
-#check t[∀α. myType %α]
-
-/--
-info: forAll ["α"]
-  (LMonoTy.tcons "arrow" [LMonoTy.ftvar "α", LMonoTy.tcons "arrow" [LMonoTy.ftvar "α", LMonoTy.tcons "int" []]]) : LTy
--/
-#guard_msgs in
-#check t[∀α. %α → %α → int]
+-- Syntax tests moved to StrataTest/DL/Lambda/LTyTests.lean
```
Comment (joscoh) [inline Strata/DL/Lambda/LTy.lean]: Don't think you need this twice (or probably at all)

File: StrataTest/Languages/Core/FactoryWF.lean
Comment (joscoh) [inline StrataTest/Languages/Core/FactoryWF.lean:1]: It seems strange that this file is in the test directory, but I'm not sure what the best organization is.

File: StrataTest/Languages/Python/Regex/ReParserTests.lean
Comment (joscoh) [inline StrataTest/Languages/Python/Regex/ReParserTests.lean:1]: This does not contain all of the tests formerly in `ReParser.lean`.

Comment (keyboardDrummer) [reply StrataTest/Languages/Python/Regex/ReParserTests.lean:1]: Oof. Did you happen to check that for the other files as well? Should I double check or do you feel you covered it?

File: StrataTest/Languages/Core/FactoryWF.lean
Comment (keyboardDrummer) [reply StrataTest/Languages/Core/FactoryWF.lean:1]: I'm not particularly inclined to also introduce a StrataProofs folder - it seems like more boilerplate without a benefit - but I'm not strongly against it. What do you think about doing this?

File: StrataTest/Languages/Python/Regex/ReParserTests.lean
Comment (joscoh) [reply StrataTest/Languages/Python/Regex/ReParserTests.lean:1]: I tried to briefly check but I cannot guarantee that I caught all of them. I noticed that that one file was particularly short compared to the original tests.

File: Strata/DDM/Util/Ion/Serialize.lean [hunk 2]
```diff
@@ -189,17 +189,14 @@ def bytesRequired (x : Nat) : Nat := aux 0 x
 Return the number of bytes using the 7-bit varint encoding.
 -/
 @[specialize]
-def varbytesRequired (x : Nat) : Nat := aux 0 x
+public def varbytesRequired (x : Nat) : Nat := aux 0 x
```
Comment (aqjune-aws) [inline Strata/DDM/Util/Ion/Serialize.lean]: I think it is not a good idea to change access modifiers due to the test support, partly because one motivation behind Lean4 module system is not to make too many definitions visible outside.
Should we keep the tests of private definitions here?

Comment (joscoh) [reply Strata/DDM/Util/Ion/Serialize.lean]: I somewhat disagree, I would rather we have a better-organized repo where tests/implementation/proofs are clearly separated even if this clashes somewhat with Lean's privacy modifiers. However, I do agree that this is not ideal, and I am not strongly opposed to changing this.

Relatedly, does Lean's module system have a way to open a particular module in a context? So we can have module `A` with an internal definition, but within `ATest`, we can `open` or `import` `A` and have access to its definitions? Seems like this should possible.

Comment (aqjune-aws) [reply Strata/DDM/Util/Ion/Serialize.lean]: > I somewhat disagree, I would rather we have a better-organized repo where tests/implementation/proofs are clearly separated even if this clashes somewhat with Lean's privacy modifiers. However, I do agree that this is not ideal, and I am not strongly opposed to changing this.

Yeah, it will be the ideal case. But I think this is indeed pros and cons in hiding the definitions that don't need to be exposed to public. Having more private definitions will make lake faster (at least in theory because it was their motivation).

> Relatedly, does Lean's module system have a way to open a particular module in a context? So we can have module A with an internal definition, but within ATest, we can open or import A and have access to its definitions? Seems like this should possible.

There is the `@[expose]` attribute that can be attached to a definition so that the downstream modules can unfold the definition. But from the user side, not clear. :/ If there isn't one indeed, I guess people will request it soon and Zulip will have interesting discussions about this.

Comment (keyboardDrummer) [reply Strata/DDM/Util/Ion/Serialize.lean]: When importing a module, `import all` can be used to also import its private definitions. In our current codebase most files aren't modules, so we can't always use this. For the sake of this PR, I ask that we allow removing some private modifiers. I will see if in a separate PR we can move everything to modules, which will allow keeping the symbols that this PR made public, private again. Since in modules everything is private by default, I think we don't need to track the specific symbols.

File: StrataTest/DL/Lambda/LExprTests.lean [hunk 3]
```diff
@@ -0,0 +1,205 @@
+/-
```
Comment (aqjune-aws) [inline StrataTest/DL/Lambda/LExprTests.lean:1]: It seems this file name does not match the pattern and should be `LExprTests.lean`. Let's avoid introducing corner cases in the pattern "{file name} + Tests.lean" for test files (modulo *Correct.lean and *WF.lean).
